src/algebra/monoid_algebra.lean

```diff
@@ -1170,6 +1170,44 @@ See note [partially-applied ext lemmas]. -/
   f = g :=
 ring_hom_ext (ring_hom.congr_fun h₁) (monoid_hom.congr_fun h_of)
 
+section opposite
+
+variables {I : Type*} [semiring R]
+
+open finsupp opposite
+
+-- simps doesn't do well here, but we could add some lemmas
+-- about `support` and `coe_fn` that are true by rfl
+/-- The opposite of an `add_monoid_algebra R I` is additively equivalent to
+the `add_monoid_algebra Rᵒᵖ I` over the opposite ring, taking elements to their opposite. -/
+protected noncomputable def op_add_equiv :
+  (add_monoid_algebra R I)ᵒᵖ ≃+ add_monoid_algebra Rᵒᵖ I :=
+opposite.op_add_equiv.symm.trans (finsupp.map_range.add_equiv opposite.op_add_equiv)
+
+@[simp]
+lemma op_add_equiv_single (i : I) (r : R):
+  add_monoid_algebra.op_add_equiv (op (single i r)) = single i (op r) :=
+by simp [add_monoid_algebra.op_add_equiv]
+
+/-- The opposite of an `add_monoid_algebra R I` is ring equivalent to
+the `add_monoid_algebra Rᵒᵖ I` over the opposite ring, taking elements to their opposite.
+For the case where the index type `I` is not `[add_comm_monoid I]`, see
+`add_monoid_algebra.op_add_equiv`. -/
+protected noncomputable def op_ring_equiv [add_comm_monoid I] :
+  (add_monoid_algebra R I)ᵒᵖ ≃+* add_monoid_algebra Rᵒᵖ I :=
+{ map_mul' := begin
+    dsimp only [add_equiv.to_fun_eq_coe, ←add_equiv.coe_to_add_monoid_hom],
+    rw add_monoid_hom.map_mul_iff,
+    ext,
+    dsimp,
+    rw ←op_mul,
+    simp only [add_monoid_algebra.single_mul_single, add_monoid_algebra.op_add_equiv_single,
+      ←op_mul, add_comm],
+  end,
+  ..add_monoid_algebra.op_add_equiv }
```

Comment:
You can actually inline everything into this proof:
```
/-- The opposite of an `add_monoid_algebra R I` is ring equivalent to
the `add_monoid_algebra Rᵒᵖ I` over the opposite ring, taking elements to their opposite.
For the case where the index type `I` is not `[add_comm_monoid I]`, see
`add_monoid_algebra.op_add_equiv`. -/
@[simps? apply {simp_rhs := tt}]
protected noncomputable def op_ring_equiv [add_comm_monoid I] :
  (add_monoid_algebra R I)ᵒᵖ ≃+* add_monoid_algebra Rᵒᵖ I :=
{ map_mul' := begin
    dsimp only [add_equiv.to_fun_eq_coe, ←add_equiv.coe_to_add_monoid_hom],
    rw add_monoid_hom.map_mul_iff,
    ext i₁ r₁ i₂ r₂ : 6,
    dsimp,
    simp only [map_range_single, single_mul_single, ←op_mul, add_comm],
  end,
  .. op_add_equiv.symm.trans $ finsupp.map_range.add_equiv (op_add_equiv : R ≃+ _) }
```

Note that `ext ... : 6` here prevents ext making things more complex than needed.